src/handler-context/make-incoming-messages.ts

```diff
@@ -78,27 +76,89 @@ const isJsonObject = (x: unknown): x is Record<string, unknown> =>
 const isSqsRecord = (x: unknown): x is { messageId: unknown } =>
   isJsonObject(x) && "messageId" in x && !!x.messageId;
 
+const SQSMessageIncludesS3EventRecord = (result: Result<any>): boolean => {
+  for (const incomingMessage of result.incomingMessages) {
+    const { body } = incomingMessage;
+    if (body.Records) {
+      for (const record of body.Records) {
+        if (record.s3) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+};
+
 export const makeIncomingMessages = async <TBody extends HandlerMessageBody>(
   event: S3Event | SQSEvent,
   incomingMessageBodyTypeGuard: (maybeBody: any) => maybeBody is TBody,
   logger: Logger,
   failuresAllowed?: boolean
-): Promise<{
-  incomingMessages: Array<HandlerIncomingMessage<TBody>>;
-  failedIds: string[];
-}> => {
-  const result = isSQSEvent(event)
-    ? makeSQSMessages(event, incomingMessageBodyTypeGuard, logger)
-    : {
-        incomingMessages: await makeS3Messages(
-          event,
-          incomingMessageBodyTypeGuard
-        ),
-        failedIds: [], // S3 events have no message IDs (throw error on failure instead)
-      };
-
-  if (!failuresAllowed && result.failedIds.length > 0)
-    throw new Error(ERROR_MESSAGE_DEFAULT);
-
-  return result;
+): Promise<Result<TBody>> => {
+  let result: Result<unknown>;
+
+  if (isSQSEvent(event)) {
+    result = makeSQSMessages(event, logger);
+
+    if (SQSMessageIncludesS3EventRecord(result)) {
+      const { id, body } = result.incomingMessages[0];
+      const failedIds = [...result.failedIds];
+      const incomingMessages = [];
+      try {
+        const s3Messages = await makeS3Messages(body as S3Event);
+        const s3MessagesWithIds = s3Messages.map((s3Message) => ({
+          id,
+          body: s3Message.body,
+          meta: s3Message.meta,
+        }));
+        incomingMessages.push(...s3MessagesWithIds);
+      } catch (error) {
+        logger.error(ERROR_MESSAGE_DEFAULT, { error, messageId: id });
+        if (id !== undefined && !failedIds.includes(id)) failedIds.push(id);
+      }
+
+      result = { incomingMessages, failedIds };
+    }
+
+    for (const incomingMessage of result.incomingMessages) {
+      try {
+        const bodyIsExpectedType = incomingMessageBodyTypeGuard(
+          incomingMessage.body
+        );
+        if (!bodyIsExpectedType) {
+          result.incomingMessages.pop();
```

Comment:
I guess this has been working because there has always been one incomingMessage in both the transaction flow and email-processing flow both when running the integration tests and manually tests. Maybe we can grab some time together to chat through this?